### PR TITLE
Add vjepa2_doctor for checkpoint/config validation before continued pretrain

### LINKED ISSUE #163: [VJEPA2.1] Crop size mismatch between uploaded checkpoint (384) and cooldown config (256) + NaN loss during continuous pretraining
## Environment
* Repository: facebookresearch/vjepa2
* Model: VJEPA2.1 (ViT-L)
* Hardware: 2× H200 NVL GPUs
* Dataset (test): ~22k videos (custom domain)

## Bug: crop size mismatch between checkpoint and cooldown config

The checkpoint uploaded by Meta for VJEPA2.1 was trained with a crop size of `384`, but the accompanying cooldown config file specifies a crop size of `256`.

### Affected files
- Checkpoint: trained and saved at resolution 384×384
- Cooldown config: `crop_size: 256`

Trying to reproduce the 2-step pretraining of VJEPA2.1 using the provided configs may lead to a mismatch with the released checkpoint. Moreover, would loading the 384-pretrained checkpoint and continuing training with the cooldown config as-is cause a positional embedding shape mismatch, or silently train at the wrong resolution?

### Context

I ran continuous pretraining of VJEPA2 (ViT-L) on a custom dataset and measured an accuracy improvement when comparing a fine-tuned linear probe on the base VJEPA2 checkpoint vs the same probe on my domain-pretrained VJEPA2 checkpoint. Given this result, I wanted to replicate the approach with VJEPA2.1 hoping for an even stronger encoder baseline. Unfortunately the run diverges immediately.

### Observed behaviour

* Loss: NaN after the first few iterations.
* Dataset size (test run): ~22k videos.
* Target dataset size (full run): ~500k videos.

### What I have tried

- Checked for `inf` / `NaN` in the input tensors before the forward pass.
- Lowered the learning rate significantly.
- Suspected the crop size mismatch as a potential cause (positional embedding interpolation artefacts).

### Questions

* Are the pretrain and cooldown config files intended to be used with the released checkpoint or as a standalone stage-2 config from scratch? If the former, can the crop size be corrected to 384?
* Are there known instability issues when continuously pretraining VJEPA2.1 on small datasets (~22k videos)?
* Is there a recommended starting LR / warmup schedule for continuous pretraining with the released VJEPA2.1 checkpoint?
* Does the crop size mismatch (384 checkpoint --> 256 config) cause positional embedding interpolation that could lead to NaN gradients?

### Additional info

The small-dataset test run (~22k videos) is intentional: I want to validate the approach before scaling to the full ~500k video corpus.

### PR DESCRIPTION
**Pull request by Abhishek21g**
## Problem

Continued pretraining from released V-JEPA 2.1 distilled checkpoints can fail silently when the training YAML does not match checkpoint metadata — e.g. crop_size 384 in the checkpoint vs 256 in cooldown config, or pred_depth 12 in checkpoint vs 24 in config (#163, #149).

## Repro

1. Download `vjepa2_1_vitl_dist_vitG_384.pt`
2. Use `configs/train_2_1/vitl16/cooldown-256px-64f.yaml` as-is
3. `crop_size` and `pred_depth` do not match the released weights → partial load / wrong resolution / NaN loss

## Fix

- `scripts/vjepa2_doctor.py` — inspects checkpoint top-level keys, infers `crop_size` from encoder `pos_embed`, counts predictor blocks, compares to YAML
- README section with usage
- `tests/scripts/test_vjepa2_doctor.py` — synthetic checkpoint fixtures

## Verification

```bash
python -m pytest tests/scripts/test_vjepa2_doctor.py -v
python scripts/vjepa2_doctor.py --checkpoint <path>.pt --config configs/train_2_1/vitl16/cooldown-256px-64f.yaml
```

Closes #163

### CODE REVIEW

**Review by Anshler — COMMENTED** [review]
## Code Review — PR #172: `vjepa2_doctor`

Thanks for this — it addresses a real footgun (#163). Overall the structure is clean, but I spotted a few things worth fixing before merge:

### :warning: Security — `torch.load(weights_only=False)`

https://github.com/facebookresearch/vjepa2/pull/172/files#diff-ae4247eed88c17b80683111a44f273e7a4a3dcfb9b3738e347228121cf88618cR86

`torch.load(path, map_location="cpu", weights_only=False)` enables arbitrary code execution via pickle. While a checkpoint inspector inherently needs to deserialize, this tool accepts a user-supplied `--checkpoint` path and could be run in CI/CD against untrusted uploads. Consider:

1. Adding a **prominent security warning** in the `--help` text and README
2. Checking the file extension / magic bytes before loading (`.pt`/`.pth`) 
3. If feasible, using `torch.load(..., weights_only=True)` and catching the error with a friendly message if the checkpoint contains custom classes

### :bug: Bug — `or` swallows falsy-zero values

https://github.com/facebookresearch/vjepa2/pull/172/files#diff-ae4247eed88c17b80683111a44f273e7a4a3dcfb9b3738e347228121cf88618cR107-R109

```python
expected_crop = _nested_get(cfg, "data", "crop_size") or cfg.get("crop_size")
expected_depth = _nested_get(cfg, "model", "pred_depth") or cfg.get("pred_depth")
expected_embed = _nested_get(cfg, "model", "embed_dim") or cfg.get("embed_dim")
```

`or` treats `0` as falsy, so a config with `crop_size: 0` or `pred_depth: 0` would fall through to the secondary lookup. In practice these values are never 0, but the pattern is fragile. Use an explicit `is not None` check:

```python
expected_crop = _nested_get(cfg, "data", "crop_size")
if expected_crop is None:
    expected_crop = cfg.get("crop_size")
```

### :mag: Edge case — `infer_crop_size` returns *last* match, not *first*

https://github.com/facebookresearch/vjepa2/pull/172/files#diff-ae4247eed88c17b80683111a44f273e7a4a3dcfb9b3738e347228121cf88618cR67-R75

The function loops over ALL keys matching `"pos_embed"`. If multiple sub-dicts exist (e.g., both encoder and predictor happen to have pos_embeds), it returns the **last** one that passes the square check. Consider returning on the **first** valid match, or limiting the search to known encoder keys.

### :white_check_mark: Missing test — `infer_embed_dim`

`infer_embed_dim` is imported in the test file but has no dedicated test. A one-liner test would ensure the shape logic is correct.

### :bulb: Suggestion — `int()` conversion in gate checks

https://github.com/facebookresearch/vjepa2/pull/172/files#diff-ae4247eed88c17b80683111a44f273e7a4a3dcfb9b3738e347228121cf88618cR113-R117

`int()` will raise `ValueError` if the config value is a non-numeric string (e.g., a YAML anchor or `null` cast). The `_nested_get` default and `or` do protect against `None`, but a bogus string like `"384px"` would crash. Consider wrapping in a try/except or validating the type first.

---

**Verdict**: Approve with minor requests — the security note on `weights_only=False` is the most important, the `or`/falsy bug and test gap are nice-to-fix.

**Comment by Abhishek21g** [comment]
Thanks @Anshler — pushed review fixes in da498da:

- **Security:** `weights_only=True` first with warned fallback; `.pt/.pth/.ckpt` + header check; security note in `--help` epilog + README
- **Bug:** `_get_config_value()` uses explicit `is None` (preserves `crop_size: 0` / `pred_depth: 0`)
- **Edge case:** `infer_crop_size` returns on first valid `pos_embed` match
- **Tests:** `infer_embed_dim`, first-match crop, zero config values, non-numeric crop string
- **int():** `_safe_int()` with `ValueError` on bogus strings like `"384px"`

Verification: `python -m pytest tests/scripts/test_vjepa2_doctor.py -v`

**Comment by Anshler** [comment]
Hi @Abhishek21g 

Sorry for bothering but it seems my Claude agent got invoked into reviewing this due to a hook in my local claude setting for this repo. I'm happy if the comment helps but just to let you know.

Thanks
